sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs

```diff
@@ -406,6 +410,250 @@ protected EventProcessor()
+
+                            // Do not bubble the exception, as the consumer is being refreshed; failure to close this consumer is non-fatal.
+                        }
+                    }
+                }
+
+                // If there was an exception captured, then surface it.  Otherwise signal that cancellation took place.
+
+                if (capturedException != null)
+                {
+                    ExceptionDispatchInfo.Capture(capturedException).Throw();
+                }
+
+                throw new TaskCanceledException();
+            }
+
+            // Start processing in the background and return the processor
+            // metadata.
+
+            return new PartitionProcessor
+            (
+                Task.Run(performProcessing, cancellationSource.Token),
+                readLastEnquedEventInformation,
+                cancellationSource
+            );
+        }
+
+        /// <summary>
+        ///   Performs the tasks needed to process a batch of events.
+        /// </summary>
+        ///
+        /// <param name="partition">The Event Hub partition whose processing should be started.</param>
+        /// <param name="eventBatch">The batch of events to process.</param>
+        /// <param name="dispatchEmptyBatches"><c>true</c> if empty batches should be dispatched to the handler; otherwise, <c>false</c>.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the processing.</param>
+        ///
+        internal virtual async Task ProcessEventBatchAsync(TPartition partition,
+                                                           IEnumerable<EventData> eventBatch,
+                                                           bool dispatchEmptyBatches,
+                                                           CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+            eventBatch = (eventBatch as IList<EventData>) ?? eventBatch.ToList();
+
+            // If there were no events in the batch and empty batches should not be emitted,
+            // take no further action.
+
+            if ((!eventBatch.Any()) && (!dispatchEmptyBatches))
+            {
+                return;
+            }
+
+            // Create the diagnostics scope used for distributed tracing and instrument the events in the batch.
+
+            using var diagnosticScope = EventDataInstrumentation.ScopeFactory.CreateScope(DiagnosticProperty.EventProcessorProcessingActivityName);
+            diagnosticScope.AddAttribute(DiagnosticProperty.KindAttribute, DiagnosticProperty.ConsumerKind);
+            diagnosticScope.AddAttribute(DiagnosticProperty.EventHubAttribute, EventHubName);
+            diagnosticScope.AddAttribute(DiagnosticProperty.EndpointAttribute, FullyQualifiedNamespace);
+
+            string diagnosticId;
```

Comment:
Sure;  was trying to avoid allocations, but I'll fully admit that there is very likely a compiler optimization already for this pattern that I'm just not aware of.  :smile: